Readme.md
# retext-porter-stemmer [![Build Status](https://img.shields.io/travis/wooorm/retext-porter-stemmer.svg?style=flat)](https://travis-ci.org/wooorm/retext-porter-stemmer) [![Coverage Status](https://img.shields.io/coveralls/wooorm/retext-porter-stemmer.svg?style=flat)](https://coveralls.io/r/wooorm/retext-porter-stemmer?branch=master)

**[retext](https://github.com/wooorm/retext)** implementation of the [Porter stemming algorithm](http://tartarus.org/martin/PorterStemmer/).

## Installation

[npm](https://docs.npmjs.com/cli/install):

```bash
$ npm install retext-porter-stemmer
```

[Component.js](https://github.com/componentjs/component):

```bash
$ component install wooorm/retext-porter-stemmer
```

[Bower](http://bower.io/#install-packages):

```bash
$ bower install retext-porter-stemmer
```

[Duo](http://duojs.org/#getting-started):

```javascript
var porterStemmer = require('wooorm/retext-porter-stemmer');
```

## Usage

```javascript
var Retext = require('retext');
var visit = require('retext-visit');
var inspect = require('retext-inspect');
var porterStemmer = require('retext-porter-stemmer');

var retext = new Retext()
    .use(inspect)
    .use(visit)
    .use(porterStemmer)

retext.parse('A simple english sentence.', function (err, tree) {
    tree.visit(tree.WORD_NODE, function (node) {
        console.log(node);
    });
    /**
     * WordNode[1] [data={"stem":"a"}]
     * └─ TextNode: 'A'
     * WordNode[1] [data={"stem":"simpl"}]
     * └─ TextNode: 'simple'
     * WordNode[1] [data={"stem":"english"}]
     * └─ TextNode: 'english'
     * WordNode[1] [data={"stem":"sentenc"}]
     * └─ TextNode: 'sentence'
     */
});
```

## API

None, **retext-porter-stemmer** automatically detects the stem of each [`WordNode`](https://github.com/wooorm/textom#textomwordnode-nlcstwordnode) (using **[wooorm/stemmer](https://github.com/wooorm/stemmer)**), and stores the stem in `node.data.stem`.


## Performance

On a MacBook Air, **retext** performs about 11% slower with **retext-porter-stemmer**.

```text
           retext w/o retext-porter-stemmer
  221 op/s » A paragraph (5 sentences, 100 words)
   24 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)

           retext w/ retext-porter-stemmer
  197 op/s » A paragraph (5 sentences, 100 words)
   20 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)
```

## Related

- [retext-lancaster-stemmer](https://github.com/wooorm/retext-lancaster-stemmer) — Same workings, but uses the Lancaster (Paice/Husk) stemming algorithm.

## License

[MIT](LICENSE) © [Titus Wormer](http://wooorm.com)


History.md

0.2.5 / 2015-01-19
==================

  * Update copyright notice in `LICENSE` to include 2015
  * Add link to whole license in `Readme.md`
  * Add Duo as an instalation method in `Readme.md`
  * Add links to installation methods in `Readme.md`
  * Refactor fences code blocks in `Readme.md`
  * Refactor to adhere to strict jsdoc style
  * Add jscs-jsdoc configuration to `.jscs.json`
  * Add jscs-jsdoc as a dev-dependency
  * Refactor npm scripts for changes in npm
  * Update eslint
  * Refactor to use generic `changeinside` in new retext
  * Update retext to stable release

0.2.4 / 2014-12-13
==================

 * Add benchmark
 * Fix wording in `Readme.md`
 * Add link to personal website to copyright in `Readme.md`
 * Update eslint, retext

0.2.3 / 2014-11-21
==================

 * Refactor `Readme.md`
 * Add flat badges to `Readme.md`
 * Fix incorrect executive rights on `test.js`
 * Refactor npm script targets in `package.json`
 * Add retext-inspect to dev-dependecies
 * Update retext

0.2.2 / 2014-10-27
==================

 * Add `.eslintrc`
 * Refactor to disallow space after object keys
 * Refactor for changes in retext
 * Update eslint, mocha, retext, retext-visit, retext-content

0.2.1 / 2014-10-21
==================

 * Refactor .jscs.json
 * Refactor test
 * Update retext

0.2.0 / 2014-10-14
==================

 * Update tests for changes in retext-visit
 * Update stemmer, retext, retext-content, retext-visit
 * Update test for changes in `stemmer`
 * Update .gitignore, .npmignore, bower ignore
 * Move spec to test.js
 * Fix reference to project name in Readme.md
 * Add missing `require` statements to examples in Readme.md

0.1.1 / 2014-09-25
==================

 * Add retext-content as a dev-dependency
 * Update Readme.md
 * Refactor API
 * Add npm deployment to travis
 * Remove `npm update` from travis
 * Remove browser test
 * Update .gitignore, .npmignore
 * Update Installation in docs
 * Add bower.json
 * Fix property order in component.json, package.json
 * Update copyright in Readme.md
 * Remove testling
 * Refactor spec for retext@0.2.0
 * Remove `stemmer` from exports
 * Update eslint, mocha, retext; broaden version ranges
 * Add section on related projects
 * Add components to gitignore

0.1.0 / 2014-07-13
==================

 * Added changelog
 * Fixed for API changes in retext
 * Update stemmer, eslint, istanbul, retext, retext-visit
 * Removed functionality to browserify, complexity unit tests by default
 * Mention which stemming implementation is used

0.0.7 / 2014-06-24
==================

5b7cae6 Updated dependency version of jscs
9f575a9 Added unit tests for 54173ce
54173ce Export the porterStemmer algorithm itself

0.0.6 / 2014-06-19
==================

e8c37e7 Unit tests can now run in the browser (spec/index.html)
922825b Fixed a bug where everything in ./spec/ was tested
d07fa43 Split the lint task into lint-api, lint-test, and lint-style
14eb0fa Updated dependency versions of stemmer, jscs, retext, and retext-visit

0.0.5 / 2014-06-17
==================

81c4409 Fix code styleguide for latest jscs updates
c967b10 Updated dependency versions of stemmer, jscs, retext, and retext-visit
a8772ab Update retext developer dependency to 0.0.10

0.0.4 / 2014-06-14
==================

4ade2c4 Added an “install with component” guide
53a1973 Fixed a bug and renamed porterStemmer to stemmer in the example
8357692 Testling badge & browser support
b0c0d98 SVG badges & coverage status badge
ac3b6be Fixed some new-lines before EOF
0eb5427 Now using wooorm/stemmer in API, removed stemmer unit tests, made everything ESLint compliant
9b264e0 Added Testling to package
f39cb5e Refactored travis-ci, added coveralls
714de54 Added stemmer as a dependency
bf9cb2f Added component.json
13f6fd2 Removed Makefile, instead opting for just package.json
994335a Replaced JSHint as a dependency with ESLint
67fdce8 Updated dependency versions of istanbul, retext, and retext visit
5eb0b52 `-` is now an allowed left-sticked operator (e.g., `-1`)

0.0.3 / 2014-06-05
==================

5b82d22 Moved the article “the” outside of a link
ecf5c18 Removed Git install “guide”
f0e2b85 Updated dependency versions of retext and retext-visit

0.0.2 / 2014-06-04
==================

6a3993d Added retext-visit as a dev dependency

0.0.1 / 2014-06-04
==================


test.js
'use strict';

/*
 * Dependencies.
 */

var Retext,
    stemmer,
    inspect,
    content,
    visit,
    assert;

Retext = require('retext');
stemmer = require('./');
inspect = require('retext-inspect');
content = require('retext-content');
visit = require('retext-visit');
assert = require('assert');

/*
 * Retext.
 */

var retext;

retext = new Retext()
    .use(inspect)
    .use(visit)
    .use(content)
    .use(stemmer);

/*
 * Fixtures.
 */

var otherWords,
    otherStems;

otherWords = ['An', 'easy', 'normal', 'paragraph'];
otherStems = ['an', 'easi', 'normal', 'paragraph'];

/*
 * Tests.
 */

describe('porterStemmer()', function () {
    it('should be a `function`', function () {
        assert(typeof stemmer === 'function');
    });

    retext.parse('A simple, english, sentence', function (err, tree) {
        it('should not throw', function (done) {
            done(err);
        });

        it('should `stem` each `WordNode`', function () {
            tree.visit(tree.WORD_NODE, function (wordNode) {
                assert('stem' in wordNode.data);
            });
        });

        it('should set each stem to `null` when a WordNode (no longer?) ' +
            'has a value', function () {
                tree.visit(tree.WORD_NODE, function (wordNode) {
                    wordNode.removeContent();

                    assert(wordNode.data.stem === null);
                });
            }
        );

        it('should automatically re-stem `WordNode`s when their values ' +
            'change',
            function () {
                var index;

                index = -1;

                tree.visit(tree.WORD_NODE, function (wordNode) {
                    wordNode.replaceContent(otherWords[++index]);

                    assert(wordNode.data.stem === otherStems[index]);
                });
            }
        );
    });
});


spec/retext-porter-stemmer.spec.js
'use strict';

var stemmer,
    content,
    visit,
    Retext,
    assert,
    retext,
    otherWords,
    otherStems;

/**
 * Module dependencies.
 */

stemmer = require('..');
Retext = require('retext');
visit = require('retext-visit');
content = require('retext-content');
assert = require('assert');

/**
 * Retext.
 */

retext = new Retext()
    .use(visit)
    .use(content)
    .use(stemmer);

/**
 * Constants.
 */

otherWords = ['An', 'easy', 'normal', 'paragraph'];
otherStems = ['An', 'easi', 'normal', 'paragraph'];

/**
 * Tests.
 */

describe('porterStemmer()', function () {
    it('should be a `function`', function () {
        assert(typeof stemmer === 'function');
    });

    it('should export an `attach` method', function () {
        assert(typeof stemmer.attach === 'function');
    });

    retext.parse('A simple, english, sentence', function (err, tree) {
        it('should not throw', function (done) {
            done(err);
        });

        it('should `stem` each `WordNode`', function () {
            tree.visitType(tree.WORD_NODE, function (wordNode) {
                assert('stem' in wordNode.data);
            });
        });

        it('should set each stem to `null` when a WordNode (no longer?) ' +
            'has a value', function () {
                tree.visitType(tree.WORD_NODE, function (wordNode) {
                    wordNode.removeContent();

                    assert(wordNode.data.stem === null);
                });
            }
        );

        it('should automatically re-stem `WordNode`s when their values ' +
            'change',
            function () {
                var index;

                index = -1;

                tree.visitType(tree.WORD_NODE, function (wordNode) {
                    wordNode.replaceContent(otherWords[++index]);

                    assert(wordNode.data.stem === otherStems[index]);
                });
            }
        );
    });
});


index.js
/**
 * Dependencies.
 */

var Retext = require('wooorm/retext@0.5.0');
var stemmer = require('wooorm/retext-porter-stemmer@0.2.5');
var dom = require('wooorm/retext-dom@0.3.2');
var visit = require('wooorm/retext-visit@0.2.5');

/**
 * Retext.
 */

var retext = new Retext()
    .use(visit)
    .use(dom)
    .use(stemmer);

/**
 * DOM elements.
 */

var $input = document.getElementsByTagName('textarea')[0];
var $output = document.getElementsByTagName('div')[0];

/**
 * Get a color representation from a string.
 */

var style = document.styleSheets[0];

var stems = {};

function hashCode(str) {
    var hash = 0;
    for (var i = 0; i < str.length; i++) {
       hash = str.charCodeAt(i) + ((hash << 5) - hash);
    }
    return hash;
}

function intToARGB(i){
    return ((i>>24)&0xFF).toString(16) +
           ((i>>16)&0xFF).toString(16) +
           ((i>>8)&0xFF).toString(16) +
           (i&0xFF).toString(16);
}

function getColorFromString(value) {
    value = intToARGB(hashCode(value)).slice(0, 6);

    while (value.length < 6) {
        value = '0' + value;
    }

    return '#' + value;
}

/**
 * Add a CSS rule.
 */

function addCSSRule(sheet, selector, rules) {
    if(sheet.insertRule) {
        sheet.insertRule(selector + '{' + rules + '}');
    } else {
        sheet.addRule(selector, rules);
    }
}

/**
 * Callback when new stems are calculated.
 */

function onstem(stem) {
    var color;

    if (stem in stems) {
        return;
    }

    color = getColorFromString(stem)
    stems[stem] = color;

    addCSSRule(style, '[title="' + stem + '"]', 'color:' + color);
}

/**
 * Events
 */

var tree;

function oninputchange() {
    if (tree) {
        tree.toDOMNode().parentNode.removeChild(tree.toDOMNode());
    }

    retext.parse($input.value, function (err, root) {
        if (err) throw err;

        tree = root;

        tree.visit(function (node) {
            if (!node.DOMTagName || !node.data.stem) {
                return;
            }

            stem = node.data.stem;

            onstem(stem);

            node.toDOMNode().setAttribute('title', stem);
        });

        $output.appendChild(tree.toDOMNode());
    });
}

$input.addEventListener('input', oninputchange);

oninputchange();


benchmark.js
'use strict';

var Retext,
    porterStemmer;

/*
 * Dependencies.
 */

Retext = require('retext');
porterStemmer = require('./');

/*
 * Dependencies.
 */

var retext,
    retextWithPorterStemmer;

retext = new Retext();
retextWithPorterStemmer = new Retext().use(porterStemmer);

/*
 * Test data: A (big?) article (w/ 100 paragraphs, 500
 * sentences, 10,000 words);
 *
 * Source:
 *   http://www.gutenberg.org/files/10745/10745-h/10745-h.htm
 */

var sentence,
    paragraph,
    section;

sentence = 'Where she had stood was clear, and she was gone since Sir ' +
    'Kay does not choose to assume my quarrel.';

paragraph = 'Thou art a churlish knight to so affront a lady ' +
    'he could not sit upon his horse any longer. ' +
    'For methinks something hath befallen my lord and that he ' +
    'then, after a while, he cried out in great voice. ' +
    'For that light in the sky lieth in the south ' +
    'then Queen Helen fell down in a swoon, and lay. ' +
    'Touch me not, for I am not mortal, but Fay ' +
    'so the Lady of the Lake vanished away, everything behind. ' +
    sentence;

section = paragraph + Array(10).join('\n\n' + paragraph);

suite('retext w/o retext-porter-stemmer', function () {
    bench('A paragraph (5 sentences, 100 words)', function (done) {
        retext.parse(paragraph, done);
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function (done) {
            retext.parse(section, done);
        }
    );
});

suite('retext w/ retext-porter-stemmer', function () {
    bench('A paragraph (5 sentences, 100 words)', function (done) {
        retextWithPorterStemmer.parse(paragraph, done);
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function (done) {
            retextWithPorterStemmer.parse(section, done);
        }
    );
});
